File: use-cases/vehicle-routing-pickup-delivery/src/main/java/org/acme/vehiclerouting/solver/justification/ServiceFinishedAfterMaxEndTimeJustification.java
package org.acme.vehiclerouting.solver.justification;

import ai.timefold.solver.core.api.score.stream.ConstraintJustification;

public record ServiceFinishedAfterMaxEndTimeJustification(String visitId, long serviceFinishedDelayInMinutes,
        String description) implements ConstraintJustification {

    public ServiceFinishedAfterMaxEndTimeJustification(String visitId, long serviceFinishedDelayInMinutes) {
        this(visitId, serviceFinishedDelayInMinutes, "Visit '%s' serviced with a %s-minute delay."
                .formatted(visitId, serviceFinishedDelayInMinutes));
    }
}

File: use-cases/vehicle-routing-kotlin/src/main/kotlin/org/acme/vehiclerouting/solver/justifications/ServiceFinishedAfterMaxEndTimeJustification.kt
package org.acme.vehiclerouting.solver.justifications

import ai.timefold.solver.core.api.score.stream.ConstraintJustification

@JvmRecord
data class ServiceFinishedAfterMaxEndTimeJustification(
    val visitId: String?, val serviceFinishedDelayInMinutes: Long,
    val description: String
) : ConstraintJustification {
    constructor(visitId: String?, serviceFinishedDelayInMinutes: Long) : this(
        visitId, serviceFinishedDelayInMinutes, "Visit '%s' serviced with a %s-minute delay."
            .formatted(visitId, serviceFinishedDelayInMinutes)
    )
}